File: 2019 PAC II - AED - 0800 - Semana de Ejercicios - Ejercicio 2/Python/BST.py
#                        ***************Nodo***************
class Node:
    def __init__(self,value):
        self.value = value
        self.left = None
        self.right = None

#                            *****Arbol Binario******
class BST:
    def __init__(self):
        
        self.root = None

    def BSTSearch(self,value, typeOfValue):
            if(typeOfValue == "number"):
                return self.searchNumerically(value,self.root)
            
            else:
                return self.searchAlphabetically(value,self.root)
            


#                            *****Arbol Binario por Numero******



    def addNumerically(self,newNode):

        return self.addInnerNumerically(newNode,self.root)
    
    
    def addInnerNumerically(self,newNode, current):
        if(not self.root):
            self.root = newNode
            return True
        
        else:
            if(current.value.number > newNode.value.number):
                if(current.left):
                    return self.addInnerNumerically(newNode,current.left)
                
                else:
                    current.left = newNode
                    return True
                
            
            else:
                if(current.right):
                    return self.addInnerNumerically(newNode,current.right)
                
                else:
                    current.right = newNode
                    return True
                   
        return False
    
    
      
    def searchNumerically(self,searchValue,current):
        if(not self.root):
            return True
        
        else:
            if(current.searchvalue.number == searchValue):
                return current
            
            elif(current.searchValue.number > searchValue):
                if(current.left):
                    return self.searchNumerically(searchValue, current.left)
                
                else:
                    return False
                
            else:
                if(current.right):
                    return self.searchNumerically(searchValue, current.right)
                
                else:
                    return False
    
    def BSTDeleteNumerically(self,value, current):
        if(not self.root):
            return False
        
        elif(self.searchNumerically(value,current)):
            return False
        
        else:
            if(self.root.value.number == value):
                leftChildren = self.root.left
                rightChildren = self.root.right

                self.root = None

                if(rightChildren):
                    self.addNumerically(rightChildren)
                
                if(leftChildren):
                    self.addNumerically(leftChildren)
                
                return True
            

            elif(current.value.number > value):
                if(current.left):
                    if(current.left.value.number == value):
                        leftChildren = current.left
                        rightChildren = current.right

                        current.left = None

                        if(rightChildren):
                            self.addNumerically(rightChildren)
                        
                        if(leftChildren):
                            self.addNumerically(leftChildren)
                        
                        return True
                    
                    else:
                        return self.BSTDeleteNumerically(value, current.left)
             
            else:
                if(current.value.number > value):
                    if(current.right):
                        if(current.right.value.number == value):
                            leftChildren = current.left
                            rightChildren = current.right

                            current.right = None

                            if(rightChildren):
                                self.addNumerically(rightChildren)
                            
                            if(leftChildren):
                                self.addNumerically(leftChildren)
                            
                            return True
                        
                        else:
                            return self.BSTDeleteNumerically(value, current.right)
    
    
    
#                            *****Arbol Binario por Nombre******

  
    
    def addAlphabetically(self,newNode):

        return self.addInnerAlphabetically(newNode,self.root)
    
    
    def addInnerAlphabetically(self,newNode, current):
        if(not self.root):
            self.root = newNode
            return True
        
        else:
            if(current.value.name > newNode.value.name):
                if(current.left):
                    return self.addInnerAlphabetically(newNode,current.left)
                
                else:
                    current.left = newNode
                    return True
                
            
            else:
                if(current.right):
                    return self.addInnerAlphabetically(newNode,current.right)
                
                else:
                    current.right = newNode
                    return True
                   
        return False

    def searchAlphabetically(self,searchValue,current):
        if(not self.root):
            return True
        
        else:
            if(current.searchvalue.name == searchValue):
                return current
            
            elif(current.searchValue.name > searchValue):
                if(current.left):
                    return self.searchNumerically(searchValue, current.left)
                
                else:
                    return False
                
            else:
                if(current.right):
                    return self.searchNumerically(searchValue, current.right)
                
                else:
                    return False

File: Lab/BST.js
/***************Nodo***************/
function Node(value){
    this.value = value;

    this.left = null;
    this.right = null;
    this.parent = null;
}

/**Arbol Binario**/
function BST(){
    this.root = null;

    this.search = BSTSearch;
    this.delete = BSTDelete;
    this.print = BSTPrint;

    //Numerico
    this.addNumerically = BSTAddNumerically;
    this.searchNumerically = BSTSearchNumerically;
    this.deleteNumerically = BSTDeleteNumerically;

    //Alfabetico
    this.addAlphabetically = BSTAddAlphabetically;
    this.searchAlphabetically = BSTSearchAlphabetically;
    this.deleteAlphabetically = BSTDeleteAlphabetically;
}

    

    function BSTSearch(value, typeOfValue){
        if(typeOfValue == "number"){
            return this.searchNumerically(value);
        }
        else{
            return this.searchAlphabetically(value);
        }
    }

    function BSTDelete(value, typeOfValue){
        if(typeOfValue == "number"){
            return this.deleteNumerically(value);
        }
        else{
            return this.deleteAlphabetically(value);
        }
    }

    function BSTPrint(current = this.root, trail =""){
        if(current){
            if(current.left){
                this.print(current.left, trail);
            }
            trail = trail + current.value.name + ": " + current.value.number + "\n";

            if(current.right){
                this.print(current.right, trail);
            }
        }
    }
    
/***********Para el arbol ordenado numericamente*********************/
    function BSTAddNumerically(newNode, current = this.root){
            if(!this.root){
                this.root = newNode;
                return true;
            }

            else{
                if(current.value.number > newNode.value.number){
                    if(current.left){
                        return this.addNumerically(newNode, current.left);
                    }
                    else{
                        current.left = newNode;
                        current.left.parent = current;
                        return true;
                    }
                }
                else{
                    if(current.right){
                        return this.addNumerically(newNode, current.right);
                    }
                    else{
                        current.right = newNode;
                        current.right.parent = current;
                        return true;
                    }
                }
            }
            return false;
    }

    function BSTSearchNumerically(value, current = this.root){
        if(!this.root){
            return null;
        }
        else{
            if(current.value.number == value){
                return current;
            }
            else if(current.value.number > value){
                if(current.left){
                    return this.searchNumerically(value, current.left);
                }
                else{
                    return null;
                }
            }

            else{
                if(current.right){
                    return this.searchNumerically(value, current.right);
                }
                else{
                    return null;
                }
            }
        }
    }


    function BSTDeleteNumerically(deleteValue){
        if(deleteValue == this.root.value.number){
            left = this.root.left;
            right = this.root.right;
            this.root = null;
            if(right){
                this.addNumerically(right);
            }
            if(left){
                this.addNumerically(left);
            }
            return true;
        }
        else{
            var parent = this.searchNumerically(deleteValue);
            if(parent.right){
                if(parent.right.value.number == deleteValue){
                        left = parent.right.left;
                        right = parent.right.right;
                        parent.right = null;
                    if(right){
                        this.addNumerically(right);
                    }
                    if(left){
                        this.addNumerically(left);
                    }
                    return true;
                }
            }
            else if(parent.left){
                if(parent.left.value.number == deleteValue){
                        left = parent.left.left;
                        right = parent.left.right;
                        parent.left = null;
                    if(right){
                        this.addNumerically(right);
                    }
                    if(left){
                        this.addNumerically(left);
                    }
                    return true;
                }
            }
        }
        return false;

    }

/***********Para el arbol ordenado alfabeticamente*********************/
function BSTAddAlphabetically(newNode, current = this.root){
    if(!this.root){
        this.root = newNode;
        return true;
    }

    else{
        if(current.value.name > newNode.value.name){
            if(current.left){
                return this.addAlphabetically(newNode, current.left);
            }
            else{
                current.left = newNode;
                return true;
            }
        }
        else{
            if(current.right){
                return this.addAlphabetically(newNode, current.right);
            }
            else{
                current.right = newNode;
                return true;
            }
        }
    }
    return false;
}

function BSTSearchAlphabetically(value, current = this.root){
if(!this.root){
    return null;
}
else{
    if(current.value.name == value){
        return current;
    }
    else if(current.value.name > value){
        if(current.left){
            return this.searchAlphabetically(value, current.left);
        }
        else{
            return null;
        }
    }

    else{
        if(current.right){
            return this.searchAlphabetically(value, current.right);
        }
        else{
            return null;
        }
    }
    }
}


function BSTDeleteAlphabetically(deleteValue){
if(deleteValue == this.root.value.name){
    left = this.root.left;
    right = this.root.right;
    this.root = null;
    if(right){
        this.addAlphabetically(right);
    }
    if(left){
        this.addAlphabetically(left);
    }
    return true;
}
else{
    var parent = this.searchAlphabetically(deleteValue);
    if(parent.right){
        if(parent.right.value.name == deleteValue){
                left = parent.right.left;
                right = parent.right.right;
                parent.right = null;
            if(right){
                this.addAlphabetically(right);
            }
            if(left){
                this.addAlphabetically(left);
            }
            return true;
        }
    }
    else if(parent.left){
        if(parent.left.value.name == deleteValue){
                left = parent.left.left;
                right = parent.left.right;
                parent.left = null;
            if(right){
                this.addAlphabetically(right);
            }
            if(left){
                this.addAlphabetically(left);
            }
            return true;
        }
    }
}
return false;

}    